components/brave_wallet/browser/internal/hd_key.cc

```diff
@@ -430,6 +431,35 @@ std::vector<uint8_t> HDKey::GetPublicKeyFromX25519_XSalsa20_Poly1305() const {
   return public_key;
 }
 
+std::vector<uint8_t> HDKey::DecryptCipherFromX25519_XSalsa20_Poly1305(
+    const std::string& version,
+    const std::vector<uint8_t>& nonce,
+    const std::vector<uint8_t>& ephemeral_public_key,
+    const std::vector<uint8_t>& ciphertext) const {
+  // Only x25519-xsalsa20-poly1305 is supported by MM at the time of writing
+  if (version != "x25519-xsalsa20-poly1305")
+    return std::vector<uint8_t>();
+  if (nonce.size() != crypto_box_curve25519xsalsa20poly1305_tweet_NONCEBYTES)
+    return std::vector<uint8_t>();
+  if (ephemeral_public_key.size() !=
+      crypto_box_curve25519xsalsa20poly1305_tweet_PUBLICKEYBYTES)
+    return std::vector<uint8_t>();
+  if (private_key_.size() !=
+      crypto_box_curve25519xsalsa20poly1305_tweet_SECRETKEYBYTES)
+    return std::vector<uint8_t>();
+
+  std::vector<uint8_t> padded_ciphertext = ciphertext;
+  padded_ciphertext.insert(padded_ciphertext.begin(), 16, 0);
+  std::vector<uint8_t> padded_plaintext(ciphertext.size() + 32);
+  if (crypto_box_open(padded_plaintext.data(), padded_ciphertext.data(),
+                      padded_ciphertext.size(), nonce.data(),
+                      ephemeral_public_key.data(), private_key_.data()) != 0)
+    return std::vector<uint8_t>();
+  std::vector<uint8_t> plaintext(padded_plaintext.cbegin() + 32,
+                                 padded_plaintext.cend() - 16);
+  return plaintext;
```

Comment:
Just wanted to also show that HTML is rendered like this.  I'm going to do a commit to rename the `message` field to `unsafe_message` though since you're right that unsafe things may be rendered by different clients using this code and it's better to indicate that it's unsafe in naming. 

I think I'll also do a change to make sure it's valid text and otherwise make the string we return a hex string. I'll follow up with a commit.

<img width="325" alt="Screen Shot 2022-03-31 at 9 04 16 AM" src="https://user-images.githubusercontent.com/831718/161061533-aca25b74-93b4-43ad-8351-00a7dadb4e3e.png">

Reply:
Ok that's good to see we're making the assumption that all text is still just a string. This definitely isn't a case that's commonly looked at either, but I think your method seems like a reasonable way to solve this for the time being and we can improve it as we want to support additional content-types. For example, rendering the image directly in there seems like something that would be pretty concerning. Your approach of just hex encoding image bytes (or any bytes for that matter) seems like a pragmatic tradeoff in UX to security scale for me.